scalate/src/main/scala/org/scalatra/scalate/ScalateUrlGeneratorSupport.scala

```diff
@@ -18,16 +18,33 @@ trait ScalateUrlGeneratorSupport extends ScalateSupport {
   override protected def createTemplateEngine(config: ConfigT) = {
     val engine = super.createTemplateEngine(config)
     //    val generatorBinding = Binding("urlGenerator", classOf[UrlGeneratorSupport].getName, true)
+    this.servletContext.addGeneratedUrls(this.getClass.getName, reflectRoutes)
     val routeBindings = this.reflectRoutes.keys map (Binding(_, classOf[Route].getName))
     engine.bindings = engine.bindings ::: routeBindings.toList
     engine
   }
 
   override protected def createRenderContext(out: PrintWriter)(implicit request: HttpServletRequest, response: HttpServletResponse) = {
+    import scala.collection.JavaConverters._
     val context = super.createRenderContext(out)
-    for ((name, route) <- this.reflectRoutes)
-      context.attributes.update(name, route)
-    //    context.attributes.update("urlGenerator", UrlGenerator)
+    val servletContext = context.asInstanceOf[org.fusesource.scalate.servlet.ServletRenderContext].servletContext
+
+    for ((servletName, routes) <- servletContext.generatedUrls) {
+      for ((name, route) <- routes) {
+        //  context.attributes.update("urlGenerator", UrlGenerator)
+        val pathForCurrentRequest = request.getServletPath
+        val mappingsForServletContainingRoute = servletContext.getServletRegistration(servletName).getMappings.asScala
+        val pathForServletContainingRoute = mappingsForServletContainingRoute.headOption.getOrElse("")
+        val pathReplacingFn = { req: HttpServletRequest =>
+          route.contextPath(req)
+            .replaceFirst(
+              pathForCurrentRequest,
+              pathForServletContainingRoute.replaceFirst("/\\*", ""))
+        }
+        val routeWithOwnServletPath = route.copy(contextPath = pathReplacingFn)
+        context.attributes.update(name, routeWithOwnServletPath)
+      }
+    }
```

Comment:
Isn't it possible to use methods in `UrlGenerator` to assemble url?

Reply:
[UrlGenerator](https://github.com/scalatra/scalatra/blob/v2.6.2/core/src/main/scala/org/scalatra/UrlGenerator.scala) does assemble urls from [Routes](https://github.com/scalatra/scalatra/blob/v2.6.2/core/src/main/scala/org/scalatra/Route.scala) - the intent of the fix made to [ScalateUrlGeneratorSupport](https://github.com/scalatra/scalatra/blob/v2.6.2/scalate/src/main/scala/org/scalatra/scalate/ScalateUrlGeneratorSupport.scala) is to save Routes with the correct context paths to generate the urls from.